File: JavaScript/Project/MVVMtpl_WinJS/MVVMtpl_WinJS.Shared/Services/StorageService.js
(function (g, WinJS) {
    "use strict";

    var applicationData = Windows.Storage.ApplicationData.current; 
    var roamingSettings = applicationData.roamingSettings;
    var localStorage = g.localStorage;
    var sessionStorage = g.sessionStorage;
    var storage = Windows.Storage;
    var localFolder = storage.ApplicationData.current.localFolder;
    var fileIO = storage.FileIO;


    var StorageService = WinJS.Class.define(
        function () {},

        {

            // LocalStorage 
            local: {
                JSON: function (key, data) {
                    if (data) {
                        localStorage[key] = JSON.stringify(data);
                    } else {
                        if (!localStorage[key]) {
                            return null;
                        }
                        return JSON.parse(localStorage[key]);
                    }
                },

                string: function (key, data) {
                    if (data) {
                        localStorage[key] = data;
                    } else {
                        return localStorage[key];
                    }
                }
            },

            // SessionStorage 
            session: {
                JSON: function (key, data) {
                    if (data) {
                        sessionStorage[key] = JSON.stringify(data);
                    } else {
                        if (!sessionStorage[key]) {
                            return null;
                        }
                        return JSON.parse(sessionStorage[key]);
                    }
                },

                string: function (key, data) {
                    if (data) {
                        sessionStorage[key] = data;
                    } else {
                        return sessionStorage[key];
                    }
                }
            },

            // Roaming 
            roaming: {
                JSON: function (key, data) {
                    if (data) {
                        roamingSettings.values[key] = JSON.stringify(data);
                    } else {
                        if (!roamingSettings.values[key]) {
                            return null;
                        }
                        return JSON.parse(roamingSettings.values[key]);
                    }
                },

                string: function (key, data) {
                    if (data) {
                        roamingSettings.values[key] = data;
                    } else {
                        return roamingSettings.values[key];
                    }
                }
            },

            // Folder
            folder: {
                setAsync: function (name) {
                    return localFolder.createFolderAsync();
                },

                getAsync: function (name) {
                    return localFolder.getFolderAsync(name);
                }
            },

            // File
            file: {
                setAsync: function (name, replace) {
                    replace = replace || storage.CreationCollisionOption.replaceExisting;
                    return localFolder.createFileAsync(name, replace );
                },

                // path: Windows.Storage.KnownFolders.picturesLibrary
                getAsync: function (name) {
                    return localFolder.getFileAsync(name);
                },

                deleteAsync: function (file) {
                    return file.deleteAsync();
                },

                appendTextAsync: function (file, text) {
                    return fileIO.appendTextAsync(file, text);
                },

                writeTextAsync: function (file, text) {
                    return fileIO.writeTextAsync(file, text);
                },

                readTextAsync: function (file) {
                    return fileIO.readTextAsync(file);
                },

                writeBytesAsync: function (file, text) {
                    var buffer = getBufferFromString(text); 
                    return fileIO.writeBufferAsync(file, buffer);
                },

                readBytesAsync: function (file) {
                    return fileIO.readBufferAsync(file);
                },
                
                copyAsync: function (file, newName, path, replace) {
                    return file.copyAsync(path || localFolder, newName, replace || storage.NameCollisionOption.replaceExisting);
                }
            }
        }
    );

    WinJS.Namespace.define("Services", {
        StorageService: StorageService
    });
})(this, WinJS);

File: CSharp/Project/MVVMtpl/MVVMtpl.Shared/Services/StorageService.cs
namespace MVVMtpl.Services
{
    using System;
    using System.Threading.Tasks;
    using Windows.Storage;

    public class StorageService
    {
        public void saveSetting(string setting, string value)
        {
            var roamingData = Windows.Storage.ApplicationData.Current.RoamingSettings;
            roamingData.Values[setting] = value;
        }

        public void deleteSetting(string setting)
        {
            var roamingData = Windows.Storage.ApplicationData.Current.RoamingSettings;
            roamingData.Values[setting] = null;
        }

        public void clearSettings()
        {
            var roamingData = Windows.Storage.ApplicationData.Current.RoamingSettings;
            roamingData.Values.Clear();
        }

        public async Task saveTextFile(StorageFolder folder, string data, string key = null)
        {
            StorageFile myFile;
            if (key == null)
            {
                myFile = await folder.CreateFileAsync("myData.txt", CreationCollisionOption.OpenIfExists);
            }
            else
            {
                myFile = await folder.CreateFileAsync(key + ".txt", CreationCollisionOption.OpenIfExists);
            }

            await FileIO.WriteTextAsync(myFile, data);
        }

        public async Task saveAppendTextFile(StorageFolder folder, string data, string key = null)
        {
            StorageFile myFile;
            if (key == null)
            {
                myFile = await folder.CreateFileAsync("myData.txt", CreationCollisionOption.OpenIfExists);
            }
            else
            {
                myFile = await folder.CreateFileAsync(key + ".txt", CreationCollisionOption.OpenIfExists);
            }
            await FileIO.AppendTextAsync(myFile, data);
            await FileIO.AppendTextAsync(myFile, Environment.NewLine);
        }

        public async Task<string> loadTextFile(StorageFolder folder, string key = null)
        {
            StorageFile myFile;
            try
            {
                if (key == null)
                {
                    myFile = await folder.GetFileAsync("myData.txt");
                }
                else
                {
                    myFile = await folder.GetFileAsync(key + ".txt");
                }
            }
            catch
            {
                myFile = null;
            }

            string text = "";
            if (myFile != null)
            {
                text = await FileIO.ReadTextAsync(myFile);
            }

            return text;
        }

        public async Task clearTextFile(StorageFolder folder, string key = null)
        {
            StorageFile myFile;
            if (key == null)
            {
                myFile = await folder.CreateFileAsync("myData.txt", CreationCollisionOption.OpenIfExists);
            }
            else
            {
                myFile = await folder.CreateFileAsync(key + ".txt", CreationCollisionOption.OpenIfExists);
            }

            await FileIO.WriteTextAsync(myFile, "");
        }

        public ApplicationDataContainer getRoamingSettings()
        {
            return ApplicationData.Current.RoamingSettings;
        }

        public StorageFolder getLocalFolder()
        {
            return ApplicationData.Current.LocalFolder;
        }

        public StorageFolder getTempFolder()
        {
            return ApplicationData.Current.TemporaryFolder;
        }

        //If device does't have SD card, returns LocalFolder
        public async Task<StorageFolder> getSdFolder()
        {
            StorageFolder firstCard = getLocalFolder();
            var devices = Windows.Storage.KnownFolders.RemovableDevices;
            var sdCards = await devices.GetFoldersAsync();
            firstCard = sdCards[0];
            return firstCard;
        }

        public StorageFolder getPicturesKnownFolders()
        {
            return KnownFolders.PicturesLibrary;
        }

        public StorageFolder getVideosKnownFolders()
        {
            return KnownFolders.VideosLibrary;
        }

        public StorageFolder getMusicKnownFolders()
        {
            return KnownFolders.MusicLibrary;
        }

        public StorageFolder getDocumentsKnownFolders()
        {
            return KnownFolders.DocumentsLibrary;
        }
    }
}
